File: src/Transport/ThreadManager.zig
const std = @import("std");
const sustenet = @import("root").sustenet;

const RwLock = std.Thread.RwLock;
const ArrayList = std.ArrayList;
const Action = sustenet.events.Action;

const ThreadManager = @This();
pub var instance: ?ThreadManager = null;

main_pool_lock: RwLock = .{},
main_pool: ArrayList(*const Action(void)),
main_pool_copied: ArrayList(*const Action(void)),
execute_event: bool = false,

pub const ThreadManagerError = error{
    InstanceIsNull,
};

/// Get the singleton instance.
pub fn getInstance(allocator: std.mem.Allocator) !ThreadManager {
    if (instance == null) {
        instance = ThreadManager{
            .main_pool = ArrayList(*const Action(void)).init(allocator),
            .main_pool_copied = ArrayList(*const Action(void)).init(allocator),
            .execute_event = false,
        };
    }

    return instance.?;
}

//#region Execution Functions
/// Sets an event to be executed on the main thread.
pub fn executeOnMainThread(
    self: *ThreadManager,
    /// The event to be executed on the main thread.
    callable: *const Action(void),
) void {
    self.main_pool_lock.lock();
    defer self.main_pool_lock.unlock();

    self.main_pool.append(callable) catch |err| {
        std.log.err("Failed to append callable to main pool: {}\n", .{err});
        return;
    };
    self.execute_event = true;
}

/// Execute all code meant to run on the main thread. Should only be called from the main thread.
pub fn updateMain(self: *ThreadManager) void {
    if (self.execute_event) {
        self.main_pool_lock.lock();
        defer self.main_pool_lock.unlock();

        self.main_pool_copied.clearAndFree();
        {
            const main_pool_slice = self.main_pool.toOwnedSlice() catch |err| {
                std.log.err("Failed to get main pool slice: {}\n", .{err});
                return;
            };
            self.main_pool_copied.appendSlice(main_pool_slice) catch |err| {
                std.log.err("Failed to append main pool slice: {}\n", .{err});
                return;
            };
            self.execute_event = false;
        }

        for (self.main_pool_copied.items) |callable| {
            callable.compute();
        }
    }
}
//#endregion

//#region Memory Functions
pub fn deinit(self: *ThreadManager) void {
    self.main_pool.clearAndFree();
    self.main_pool.deinit();
    self.main_pool.clearAndFree();
    self.main_pool_copied.deinit();
    instance = null;
}
//#endregion

File: rust/src/transport/thread_manager.rs
use lazy_static::lazy_static;
use std::collections::VecDeque;
use std::sync::{ Arc, Mutex };

pub struct ThreadManager {
    main_pool: Arc<Mutex<VecDeque<Box<dyn FnOnce() + Send>>>>,
    side_pool: Arc<Mutex<VecDeque<Box<dyn FnOnce() + Send>>>>,
    execute_main_event: Arc<Mutex<bool>>,
    execute_side_event: Arc<Mutex<bool>>,
}

lazy_static! {
    static ref INSTANCE: Arc<ThreadManager> = Arc::new(ThreadManager::new());
}

impl ThreadManager {
    fn new() -> Self {
        ThreadManager {
            main_pool: Arc::new(Mutex::new(VecDeque::new())),
            side_pool: Arc::new(Mutex::new(VecDeque::new())),
            execute_main_event: Arc::new(Mutex::new(false)),
            execute_side_event: Arc::new(Mutex::new(false)),
        }
    }

    /// Get the singleton instance.
    pub fn get_instance() -> Arc<ThreadManager> {
        INSTANCE.clone()
    }
    //#region Execution Functions
    /// Sets an event to be executed on the main thread.
    /// * `callable` - The event to be executed on the main thread.
    pub fn execute_on_main_thread(&self, action: Box<dyn FnOnce() + Send>) {
        let mut execute_main_event = self.execute_main_event.lock().unwrap();
        *execute_main_event = true;

        let mut main_pool = self.main_pool.lock().unwrap();
        main_pool.push_back(action);
    }

    /// Execute all code meant to run on the main thread. Should only be called from the main thread.
    pub fn update_main(&self) {
        let mut execute_main_event = self.execute_main_event.lock().unwrap();
        if *execute_main_event {
            let mut main_pool = self.main_pool.lock().unwrap();

            *execute_main_event = false;
            drop(execute_main_event);

            let mut main_pool_copied = VecDeque::new();
            std::mem::swap(&mut *main_pool, &mut main_pool_copied);

            drop(main_pool);

            while let Some(action) = main_pool_copied.pop_front() {
                action();
            }
        }
    }

    pub fn execute_on_side_thread(&self, action: Box<dyn FnOnce() + Send>) {
        let mut execute_side_event = self.execute_side_event.lock().unwrap();
        *execute_side_event = true;

        let mut side_pool = self.side_pool.lock().unwrap();
        side_pool.push_back(action);
    }

    pub fn update_side(&self) {
        let mut execute_side_event = self.execute_side_event.lock().unwrap();
        if *execute_side_event {
            let mut side_pool = self.side_pool.lock().unwrap();

            *execute_side_event = false;
            drop(execute_side_event);

            let mut side_pool_copied = VecDeque::new();
            std::mem::swap(&mut *side_pool, &mut side_pool_copied);

            drop(side_pool);

            while let Some(action) = side_pool_copied.pop_front() {
                action();
            }
        }
    }

    pub fn deinit(&self) {
        let mut main_pool = self.main_pool.lock().unwrap();
        main_pool.clear();

        let mut side_pool = self.side_pool.lock().unwrap();
        side_pool.clear();
    }
}
